crates/red_knot_workspace/src/workspace.rs

```diff
@@ -188,35 +188,59 @@ impl Workspace {
-    pub fn check(self, db: &RootDatabase) -> Vec<String> {
+    pub fn check(self, db: &RootDatabase) -> Vec<CompileDiagnostic> {
         let workspace_span = tracing::debug_span!("check_workspace");
         let _span = workspace_span.enter();
 
         tracing::debug!("Checking workspace");
-        let files = WorkspaceFiles::new(db, self);
-        let result = Arc::new(std::sync::Mutex::new(Vec::new()));
-        let inner_result = Arc::clone(&result);
-
-        let db = db.snapshot();
-        let workspace_span = workspace_span.clone();
-
-        rayon::scope(move |scope| {
-            for file in &files {
-                let result = inner_result.clone();
-                let db = db.snapshot();
-                let workspace_span = workspace_span.clone();
-
-                scope.spawn(move |_| {
-                    let check_file_span = tracing::debug_span!(parent: &workspace_span, "check_file", file=%file.path(&db));
-                    let _entered = check_file_span.entered();
-
-                    let file_diagnostics = check_file(&db, file);
-                    result.lock().unwrap().extend(file_diagnostics);
-                });
-            }
+
+        // TODO: We first check the files concurrently and then query the accumulated diagnostics
+        // in a second step. This should give us the benefit of parallelism and the
+        // collecting of diagnostics should be fast because it's all cached. However,
+        // traversing the entire dependency tree of `check_workspace_impl` is not cheap
+        // and it could be avoided if `check_workspace_par` could be a query instead.
+        // However, this isn't possible today because:
+        // * We need to have a single root query to get all diagnostics or we risk duplicate diagnostics
+        //   e.g. when two files call the same function that has an invalid return type annotation.
+        // * `check_workspace_par` can't be a query because Salsa can't track the dependencies
+        //   when forking *inside* a query without [parallel DB](https://github.com/salsa-rs/salsa/pull/568) support.
+        //
+        // The solution with parallel-db support is to make `check_workspace_par` a salsa query,
+        // remove `check_workspace_sync` and retrieve the accumulated diagnostics from `check_workspace_par`.
+        check_workspace_par(db, self);
+
+        let mut workspace_diagnostics: Vec<_> =
+            check_workspace_sync::accumulated::<CompileDiagnostic>(db, self);
+
+        workspace_diagnostics.sort_unstable_by(|a, b| {
+            a.file().cmp(&b.file()).then_with(|| {
+                a.range()
+                    .map_or(TextSize::default(), ruff_text_size::TextRange::start)
+                    .cmp(
+                        &b.range()
+                            .map_or(TextSize::default(), ruff_text_size::TextRange::start),
+                    )
+            })
+        });
```

Comment:
I find the formatting here makes it quite hard to read. Just importing `TextRange` into the namespace would make it much more readable, in my opinion:

```diff
--- a/crates/red_knot_workspace/src/workspace.rs
+++ b/crates/red_knot_workspace/src/workspace.rs
@@ -14,6 +14,7 @@ use ruff_db::{
     system::{walk_directory::WalkState, SystemPath, SystemPathBuf},
 };
 use ruff_python_ast::{name::Name, PySourceType};
+use ruff_text_size::TextRange;
 
 use crate::db::Db;
 use crate::db::RootDatabase;
@@ -215,11 +216,8 @@ impl Workspace {
         workspace_diagnostics.sort_unstable_by(|a, b| {
             a.file().cmp(&b.file()).then_with(|| {
                 a.range()
-                    .map_or(TextSize::default(), ruff_text_size::TextRange::start)
-                    .cmp(
-                        &b.range()
-                            .map_or(TextSize::default(), ruff_text_size::TextRange::start),
-                    )
+                    .map_or(TextSize::default(), TextRange::start)
+                    .cmp(&b.range().map_or(TextSize::default(), TextRange::start))
             })
         });
```